### ISSUE TITLE
Add a local hex2bin() function to retain support for PHP 5.3

### ISSUE THREAD

**Issue by MarkMaldaba** [dompdf/dompdf#1692]
I noticed that issue #1597 pointed out that DomPDF requires PHP 5.4 due to the use of the `hex2bin()` function, and that the version number was therefore bumped accordingly.

If this is a deliberate move to drop PHP 5.3 support, then that's a shame, but understandable - in which case, please close this ticket.

However, based on #1597 it looks like the intention was to continue to support PHP 5.3, but this function slipped through.

If this is the only reason that PHP 5.4 is required then it would be simple and desirable to include a local version of the function and therefore to retain support for PHP 5.3.  For example, here is an implementation from https://github.com/mmeyer2k/retro (based on a [php.net comment](http://www.php.net/manual/en/function.hex2bin.php#110973)):

````php
if (!function_exists('hex2bin')) {
	/**
	 * Convert hex to binary.
	 * 
	 * @param string $hexstr Hexidecimal string to convert to binary.
	 * 
	 * @return string
	 */
	function hex2bin($hexstr)
	{
		$n = strlen($hexstr);
		$sbin = '';
		$i = 0;
		while ($i < $n) {
			$a = substr($hexstr, $i, 2);
			$c = pack('H*', $a);
			if ($i == 0) {
				$sbin = $c;
			} else {
				$sbin .= $c;
			}
			$i+=2;
		}
		return $sbin;
	}
}
````

**Comment by PHPGangsta** [comment]
I'm not part of the Dompdf team, just another user.

> then that's a shame

It's a shame to still use PHP 5.3 :-)
The end-of-live of PHP 5.3 was 14 Aug 2014: http://php.net/eol.php
You definitely should upgrade from time to time to a supported version, PHP 7.1 or 7.2 nowadays: http://php.net/supported-versions.php
(or at least 5.6 or 7.0, which are in "security fixes only"-state)

Don't expect libraries to not break backwards compatibility for so many years. Some libraries raise the minimum version just because people should use supported PHP versions, and push people forward. Dompdf is really conservative to still maintain support for 5.4.

7.2 is round about 3x-5x times faster than 5.3. Dompdf heavily benefits from that because it's CPU-bound.

Of cause you can still stay at PHP 5.3, but then you cannot benefit from new Dompdf features.
Hopefully that ancient PHP version is not reachable from the Internet.
Please don't blame others to move forward.

**Comment by simonberger** [comment]
The reason and version step to break compatibility to PHP 5.3 wasn't that good. But beside the fact that PHP 5.3 is really old and unsupported since quiet a long time it is a burden to work with. One always has to remember the old array syntax for example.
If you are unable to update PHP 5.3 in your environment you have to either stick to an older DomPDF version or fork it and do those little changes.
But for the near future we have to think of refactoring the code and replace all array syntax and more. At least for new features that most likely will be done.

**Comment by MarkMaldaba** [comment]
Dropping support for old versions is fine, but it needs to be managed responsibly and deliberately - it should never happen accidentally and ideally it would result in an increment of the major version number and a prominent notice in the release notes.

As I say, if this is the only thing that stops PHP 5.3 from being supported then I would recommend adding the shim to the code-base and releasing a 0.8.3 version that fixes support for PHP 5.3, and then dropping support for it properly in the next release, numbered 0.9.0 (assuming that is the intention).

Or, if this is just one of many things that already break support for PHP 5.3, then please update the release notes for previous versions so that there is a clear indication of when support was actually dropped, so that users with operational requirements for supporting PHP 5.3 are able to safely deploy the correct legacy version.

**Comment by MarkMaldaba** [comment]
> please update the release notes for previous versions so that there is a clear indication of when support was actually dropped

Is it safe to assume the release notes are correct, and that 0.8.0 supported PHP 5.3, whilst 0.81 and above require PHP 5.4.

If so, please add a not to the list of features for 0.8.1 indicating that support for PHP 5.3 was dropped in this release.

If not, please correct the version requirements so that we can reliably know which version is safe to use if 5.3 support is required.
